src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -239,25 +207,23 @@ private void ConfigureApplication(WebHostBuilderContext context, IApplicationBui
                 return _builtApplication.BuildRequestDelegate();
             });
 
-            // Implicitly call UseEndpoints() at the end of the pipeline if UseRouting() was called implicitly.
-            // We could add this to the end of _buildApplication instead if UseEndpoints() was not so picky about
-            // being called with the same IApplicationBluilder instance as UseRouting().
-            if (implicitRouting)
-            {
-                app.UseEndpoints(_ => { });
-            }
+            // We don't know if user code called UseEndpoints(), so we will call it just in case, and we will make sure we are the only ones setting
+            // the EndpointDataSource in RouteOptions with this property
+            app.Properties[GlobalEndpointBuilderCopyRoutesKey] = null;
+            app.UseEndpoints(_ => { });
+            app.Properties.Remove(GlobalEndpointBuilderCopyRoutesKey);
 
             // Copy the properties to the destination app builder
             foreach (var item in _builtApplication.Properties)
             {
                 app.Properties[item.Key] = item.Value;
             }
-        }
 
-        private static IEndpointRouteBuilder? GetEndpointRouteBuilder(IApplicationBuilder app)
-        {
-            app.Properties.TryGetValue(EndpointRouteBuilderKey, out var value);
-            return (IEndpointRouteBuilder?)value;
+            // Remove the route builder to clean up the properties, we're done adding routes to the pipeline
+            // REVIEW: this makes startup filter with userouting/useendpoints fail unless we don't remove the EndpointRouteBuilder in UseEndpoints if a global one exists
```

Comment:
I can't parse this but I'm betting is a new complication.

Reply:
Basically if a startup filter has UseRouting() it will set a route builder, then it'll call into the WebApplicationBuilder which calls UseRouting with a global route builder, and UseEndpoints which always removes the non-global route builder. Then when the filter continues after WebApplicationBuilder it calls UseEndpoints which will fail because no route builder exists.

So if you look at UseEndpoints, I currently check if a global route builder exists and don't remove the non-global route builder in that case.

Reply:
Should we only remove the local route builder if the relevant application builder is the local one? That would make the most sense right?

Reply:
Take a look at the latest changes

Reply:
I saw it!